### ISSUE TITLE
New layout templates *installed manually* require theme cache clear

### ISSUE THREAD

**Issue by docwilmot** [backdrop/backdrop-issues#3665]
**Description of the bug**
Similar to https://github.com/backdrop/backdrop-issues/issues/984
If you add a new layout template manually, then change an existing layout to use that template, Backdrop actually doesnt render the new template until a theme cache clear. Thats probably because Layout module in `layout_theme()` creates a theme item for each template.

**Steps To Reproduce**
Copy Moscone and rename it `Foo` (best to use a template with a sidebar to test this), and rename the files (`foo.info`, `layout--foo.tpl.php`). Remember to change the name in `foo.info`.
Put it in `ROOT/layouts`
Change the default layout to use Foo (ie go to `admin/structure/layouts/manage/default/configure` and click Foo then `Save layout`

**Actual behavior**
When the block add page for default layout loads once youve saved to the new template, you'll note there is no sidebar. 
Core is actually using the default legacy `layout-one-column` to render the page because there isnt a theme entry for Foo.
Clear theme cache and the new Foo layout starts being used.

**Expected behavior**
Foo should be used immediately

**Additional information**
1.12

---
~~PR by @klonos: https://github.com/backdrop/backdrop/pull/2579~~
PR by @docwilmot: https://github.com/backdrop/backdrop/pull/3107

**Comment by klonos** [comment]
PR up for review. Small change that fixes a bug, so proposing 1.12.6 for milestone.

**Comment by docwilmot** [comment]
Code seems good, and it works. Minor quibble, not worth delaying for. RTBC.

**Comment by klonos** [comment]
Quibble fixed @docwilmot 😉

**Comment by quicksketch** [comment]
I'm not sure I agree with this. Backdrop code will never clear the theme registry based on an administrative action, other than things that clearly affect functionality, like enabling a module or clearing the caches.

It's an expectation that if you're manually modifying code on the disk, then you'll have to clear caches to get some of those change to take effect. Although I don't really _like_ that expectation, this doesn't seem any different to me than if you created a new template file or preprocess function in your theme. You have to clear the cache to get those changes to be picked up. I don't think we should be making a special exception for layout templates that they flush the entire theme registry cache when nothing else has this kind of automatic flushing.

**Comment by docwilmot** [comment]
What you say makes sense. The difference here though is that the new template actually shows up in the template list, allows itself to be manipulated, but loads the wrong template tpl file. That seems like an exception, in that a new preprocess, for example, wouldn't "exist" at all before the cache clear. 

Don't know if that is enough to make an exception. Would like to hear what others think.

**Comment by BWPanda** [comment]
Just based on what @quicksketch said, I'd say that its reasonable to expect that someone who knows how to write a new template file or preprocess function would also know/figure out that they need to clear caches too. I'm not sure the same can be said of someone using the Layouts UI though...

**Comment by oadaeh** [comment]
Only based on the comments here, the two things I would add are:
* I think that adding the automatic cache clear for this one thing would lead to inconsistent behavior that would confuse some people. (Was the cache cleared here or not?)
* I think adding this has the potential for introducing unintended side effects. It might look like the code is contained, but maybe something somewhere else is expecting certain behavior that this might change.

**Comment by herbdool** [comment]
Can we add a message that caches were cleared?

**Comment by BWPanda** [comment]
Maybe a weird suggestion, but what about a setting where you can choose which events will clear caches automatically?

**Comment by klonos** [comment]
> ...I'd say that its reasonable to expect that someone who knows how to write a new template file or preprocess function would also know/figure out that they need to clear caches too.

I also see how this change here would create a "bad precedent", but it's these kinds of assumptions/expectations that people would or should know how things are done that have gotten Drupal the "steep learning curve" reputation. We have been trying hard (and are actually doing a great job) to correct this in Backdrop, so I say that if we can make things work for people, without them having to "know how things work", then we should do it.

> Backdrop code will never clear the theme registry based on an administrative action, other than things that clearly affect functionality, like enabling a module or clearing the caches.

I think that this qualifies as such a case. Same as #984. #984 was a UX WFT; I see this one here is a UX/DX WFT.

**Comment by herbdool** [comment]
Maybe a compromise is to display a message with a link to clear the cache.

**Comment by quicksketch** [comment]
A button at the top of the layout list that said something like "Refresh Layout Templates" would also be acceptable IMO.

**Comment by jenlampton** [comment]
An interesting discussion!  I agree that a developer writing code, or moving files on disk should be expected to know when to clear caches or rebuild the theme registry.

But does this affect the new flexible layout builder too? In that case we are putting the code-building tools (creating a layout) into the hands of someone who should not necessarily know when a cache needs to be cleared, or a registry rebuilt. In that case I think we should do the rebuild for them.

I like the compromise of adding a button to the top of the layout list (or maybe layout template list?) that rebuilds the registry.

**Comment by docwilmot** [comment]
> A button at the top of the layout list that said something like "Refresh Layout Templates" would also be acceptable IMO.

That would not really help in this situation IMO. Keep in mind that once you add a new template, it shows up in the list already. To the uninitiated, there would be no need to refresh the list: your new template is already there. But when you try to use it, that would fail for mysterious reasons. 

And I think adding a cache clear button when there is a cache clear admin bar link is superfluous. 

The decision here is either we automatically refresh the cache or we dont and assume users will flush it manually. THis has not shown up as an issue for the last few years so I dont thin its a major problem either way. 

Happy to close this if the auto-cache-clear breaks precedent too much.

**Comment by stpaultim** [comment]
So, just after @docwilmot notes that this has not shown up as an issue in a while, a user posts an issue in Github about this issue. https://github.com/backdrop/backdrop-issues/issues/4242

According to this user "It helped, thank you so much! I've tried clearing layouts cache, but it didn't occur to me that I need to clear themes cache."

**Comment by klonos** [comment]
Coming back to @quicksketch's objection to @docwilmot's RTBC'ing this:

> I'm not sure I agree with this. Backdrop code will never clear the theme registry based on an administrative action, other than things that clearly affect functionality, like enabling a module or clearing the caches.

Adding layout templates is similar to enabling a module, and it does add functionality (adds layout templates that are made available for selection). Rehashing @Slepana's comment from #4242:

> I will add though, while the other issue says that it has to be a manually added template, for me it occurred even with templates I installed using the backdrop administration UI..

All this + this comment by @docwilmot:

> ...your new template is already there. But when you try to use it, that would fail for mysterious reasons.

...has made me think that perhaps the culprit here is that we are using theme cache for the layout templates, when we should be using a separate cache bin. So maybe refactoring this to use a separate bin would be best(?). That way, we'd not need to clear the entire theme cache - only this specific bin. I am not familiar with how caching internals work though, so cannot implement something like that 😞

**Comment by Slepana** [comment]
I was asked to comment that this also happened to me when I installed new layouts through the UI, which is true, but I see that @klonos already copied this part here.

I also see that some people wonder what kind of user could even stumble into this kind of issue to begin with, so I'll describe my thought process in hopes it helps you find the best way to solve the issue:

1. I have no experience with Backdrop, Drupal or any CMS for that matter, but I do have Web Development experience, so I'm tend to not be afraid of touching code.
 
2. I started by playing with flexible layouts which worked well for me. However, I eventually ran into some limitations. From what I've read so far about Backdrop everything seemed to point that Themes and Layouts are completely separate compared to Drupal (which I'm not familiar with, so I wouldn't be able to recognize any leftover connections) So I _skipped the themes_, opened the layouts documentation and got to work.

3. I open up the menu and see my layout there, but it looks different than I expect. I did quickly remember that the docs mention a default file, so this wasn't very confusing.

4. I spend a considerable amount of time trying to fix my template because the way I saw it _"clearly backdrop loaded my template through some magic already, else it wouldn't be in the list"_ so I assume that I must've named it incorrectly or it's somehow broken. 

5. I try to paste the contents into a core template and it works fine. I begin to suspect that maybe it isn't actually loading, but the docs mention no extra steps (in contrast to creating themes page which has a note about this specific behavior, but one that I sadly haven't yet read at the time).

6. I decide to download a new layout through the UI, partially to see if I can spot some differences and partially hoping that whatever backdrop does to load them also loads my files. To my surprise, they don't work either. This is the part where I begin to think that this is clearly a bug, because even modules, which are much more complex, don't seem to require extra steps if you install them through backdrop itself.

7. I finally notice the "flush all caches" link, but then I see a list pop up which has layouts in it. Thinking that layouts cache must be what stores these settings I click on it but it still doesn't work. I don't know why I didn't try to flush all caches, I guess I was tired by then. Given that I was certain it must be some kind of bug by then, I didn't try anything else.

I apologize if this is too long, I wasn't sure what's significant. Given how close I got to making it work, it's not all that surprising that people didn't report it much, but it's still confusing. Maybe the note from creating themes docs could be copied to layouts? It would prevent what happened to me at least. Wouldn't help people with no coding experience just trying to download the layouts though.

**Comment by stpaultim** [comment]
@Slepana - Thanks for this. I think it's really helpful to have this kind of detailed explanation of what is going through your mind as you try a new feature. Especially from someone with fresh eyes.

**Comment by docwilmot** [comment]
Yes thanks @Slepana.

I think then this is clearly something we should not expect users to discover on their own, and is likely to be relatively common if users add new templates, either manually or through the Installer. I think we *have* to flush the theme cache on the user's behalf.

If need be we could check and see if a theme entry exists for the newly selected template, and only flush caches in that case.

**Comment by docwilmot** [comment]
Although this PR was by @klonos I'll try and come up with a solution to this.

**Comment by jenlampton** [comment]
> we could check and see if a theme entry exists for the newly selected template

The layout system works by using theme hook suggestions to replace `layout.tpl.php` with the current selected layout template, so this should always be the case.

**Comment by docwilmot** [comment]
Yes but theme entries are cached, so if you just added a template an entry wouldn't yet be in the cache, which is the whole problem. So check cache, if template exists but cache of template doesn't, flush cache

**Comment by jenlampton** [comment]
Ah that makes sense :)

**Comment by docwilmot** [comment]
New PR checks if there is a theme entry before clearing cache. We could also instead show a message directing the user to clear the cache, but I think we should do it for him automatically.

**Comment by herbdool** [comment]
@docwilmot is it possible to only build the one cache entry instead of clearing the whole cache? And delete the old entry if there was one?

**Comment by herbdool** [comment]
Though looking at the code my idea might not be the recommended way. I'm supporting @docwilmot PR. It is selective of when it clears theme registry and we need to make easier for everyone.

**Comment by BWPanda** [comment]
Waiting on @docwilmot to refresh PR to see if that fixes test errors...

**Comment by docwilmot** [comment]
All green

**Comment by BWPanda** [comment]
Thanks everyone for the discussion, particularly @docwilmot and @klonos for the PRs! I've merged https://github.com/backdrop/backdrop/pull/3107 into 1.x and 1.16.x.
